src/backend/distributed/planner/multi_planner.c

```diff
@@ -701,6 +724,16 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo, Index
 	restrictionContext = CurrentRelationRestrictionContext();
 	Assert(restrictionContext != NULL);
 
+	/*
+	 * Before reaching this point, some code blocks (for example GEQO optimizations)
+	 * may switch to their own memory context. If we allocate the memory here and the
+	 * lappend calls below in their context, allocated memory will be freed with that
+	 * code block's memory context. We need longer lifespan for these variables, so we
+	 * allocate them in the CurrentRestrictionsMemoryContext(), which is set to
+	 * CurrentMemoryContext at CreateAndPushPlannerRestrictionContext.
+	 */
+	oldMemoryContext = MemoryContextSwitchTo(restrictionsMemoryContext);
```

Comment:
Shouldn't duplicate larger comments, how about referring to the other hook instead?